### ISSUE TITLE
Feature request: support ES7 decorators

### ISSUE THREAD

**Issue by jacopotarantino** [jsdoc3/jsdoc#1165]
I'm trying to run JSDoc on a file that uses a decorator (everything is run through webpack+babel. only jsdoc runs on the raw files and generally handles es6 features in a surprisingly forward manner). the decorator, however, causes the following error message:

```bash
ERROR: Unable to parse /Users/jack/mycomponent/source/my-component.js: Line 36: Unexpected token ILLEGAL
```

The file looks like this:

```javascript
export class MyComponent extends UIComponent {
  /**
   * Creates a new MyComponent
   * @extends UIComponent
   */
  constructor () {
    super()
    // extend component...
  }

  /**
   * Puts this component on the page with a custom render method.
   */
  @deprecate('this method is no longer valid')
  render () {
    // do stuff
  }
}
```

After a bit of research I'm certain that JSDoc doesn't support decorators and whatever the problem is comes from the underlying JS engine or parser. Language transpiling and ASTs are a bit out of my scope of expertise so I'm hoping somebody can help me either write a plugin for this or somebody can tell me a really quick way to hack support for decorators into JSDoc until it becomes a core feature.

**Comment by bewl** [comment]
Is there any update on this? I am trying to incorporate JSDoc into an Aurelia application where we use decorators often.

**Comment by leeyeh** [comment]
AFAIK, decorator won't be supported shortly by espree, since it is at stage-1.  https://github.com/eslint/espree/issues/211

**Comment by sveinburne** [comment]
+1 ; It would be nice that, at least, that the decorator literal does not triggers an error. 
Perhaps through a plugin that explicitly supports [babel decorators legacy plugin](https://github.com/loganfsmyth/babel-plugin-transform-decorators-legacy)?

**Comment by damianobarbati** [comment]
any news on this...? As I found async/await was finally supported.. doh, @decorators are not! T_T

**Comment by hegemonic** [comment]
JSDoc now uses [Babylon](https://github.com/babel/babylon) as its parser, which means it's possible to add support for decorators. However, it will still take a fair amount of work to implement this feature, so it's not going to happen overnight.

Please don't add +1 comments or ask for updates—I know lots of people want this feature, and I'll get to it as soon as I can.

**Comment by hegemonic** [comment]
For now, you can use the [jsdoc-babel](https://github.com/ctumolosus/jsdoc-babel) plugin to preprocess JS files that use decorators. If you have issues with that plugin, please post them to the [jsdoc-babel repo](https://github.com/ctumolosus/jsdoc-babel/issues).

**Comment by damianobarbati** [comment]
Yep @hegemonic, I'm already using what you just said as a workaround.
Thanks for your work!

**Comment by hegemonic** [comment]
Decorators can now be parsed on GitHub master, without the need for a plugin. This change will be included in JSDoc 3.5.0.

However, JSDoc does not currently add any info about decorators to the generated documentation. (That's the part that will take some work, and that may not make it into JSDoc 3.5.0.)

**Comment by mbehzad** [comment]
for me it still fails parsing when a get method has the decorator: 
`Unable to parse ... Stage 2 decorators may only be used with a class or a class method` 
can the `member.kind` check be extended to "get" in addition to "method"?

**Comment by hegemonic** [comment]
@mbehzad That error message is coming from [Babylon](https://github.com/babel/babylon), our JavaScript parser. Please file an issue in the Babylon repo.

**Comment by mbehzad** [comment]
@hegemonic, thanx for your hint. i opened an issue [#756](https://github.com/babel/babylon/issues/756) and a PR [#766](https://github.com/babel/babylon/pull/766) in Babylon repo (it is now merged).

**Comment by jwanga** [comment]
@mbehzad for you and for posterity. The latest development version of JSDoc seems to have fixed the issue.

`npm install git+https://github.com/jsdoc3/jsdoc.git --save-dev`
